### PR TITLE
Enablement of NHWC FP16 ops (Conv + Batchnorm) on selected AMD GPUs

### PR DESCRIPTION
**Pull request by stevenireeves**
Enables FP16 NHWC in XLA and non XLA paths for convolutions and batch normalization. Currently the MIOpen NHWC Batchnorm isn't optimized and we witness an approximate 4x reduction in performance when using it. 

In order to test NHWC in the future, set the environment variable TF_USE_ROCM_NHWC=1 before running your test. 
Some data on a small model (Conv -> BN -> Dense):

NHWC
Name	Calls	Total Duration Ns	Avg Ns
MIOpenBatchNormBwdSpatial.kd	10	4996320	499632
MIOpenBatchNormFwdTrainSpatial.kd	10	2183200	218320

NCHW
Name	Calls	Total Duration Ns	Avg Ns
MIOpenBatchNormBwdSpatial.kd	10	1034400	103440
MIOpenBatchNormFwdTrainSpatial.kd	10	560959	56095

NHWC BN is ~5x slower in Backwards and ~3.9x slower in Forwards pass on MI100.

### CODE REVIEW

**Comment by stevenireeves** [comment]
jenkins : retest Ubuntu-GPU-multi please

File: tensorflow/core/kernels/pooling_ops_common.cc [hunk 1]
```diff
@@ -291,7 +291,6 @@ void DnnPoolingOp<T>::Compute(OpKernelContext* context,
                                           ToString(data_format)));
   }
 #endif
-
```
Comment (deven-amd) [inline tensorflow/core/kernels/pooling_ops_common.cc]: nit: drop this change, to remove the file from the changelist

File: tensorflow/core/kernels/fused_batch_norm_op.cc [hunk 2]
```diff
@@ -676,6 +678,15 @@ bool BatchnormSpatialPersistentEnabled() {
...
   return is_enabled;
+#elif TF_ROCM_VERSION >= 50000
+  static bool is_enabled = [] {
+    bool is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+        "TF_USE_ROCM_NHWC",
+        /*default_val=*/false, &is_enabled));
+    return is_enabled;
+  }();
+  return is_enabled;
```
Comment (deven-amd) [inline tensorflow/core/kernels/fused_batch_norm_op.cc]: Is this the correct place for this check?

What we really want is the `fast_nhwc_batch_norm` var to evaluate to `true` when this condition is met. The change above will work, but we seem to be hijacking the unrelated `BatchnormSpatialPersistentEnabled` routine for it.


How about we have something like
```
bool useNhwcLayoutForBatchNormOnRocm() {
#if TENSORFLOW_USE_ROCM && (TF_ROCM_VERSION >= 50000)
     // add above logic here...perhaps have separate env vars for conv and batchnorm
#else                 
   return false;
}

bool useNhwcLayoutForConvOnRocm() {
...
}
```
If we can stick them in a header file, which is visible to the layout optimizer, conv impl, and  batchnorm impl codes, then we call avoid the code duplication

That will also help get rid of the other `(TF_ROCM_VERSION >= 50000)` checks in this file

File: tensorflow/core/kernels/conv_ops.cc [hunk 3]
```diff
@@ -868,8 +872,18 @@ void LaunchConv2DOp<GPUDevice, T>::operator()(
   const bool compute_in_nhwc = DataTypeToEnum<T>::value == DT_HALF &&
                                stream->GetCudaComputeCapability().IsAtLeast(
                                    se::CudaComputeCapability::VOLTA);
+#elif TENSORFLOW_USE_ROCM && TF_ROCM_VERSION >= 50000
+  // AMD Matrix Cores on MI100 and MI200 allow for efficient FP16 NHWC convolutions
```
Comment (deven-amd) [inline tensorflow/core/kernels/conv_ops.cc]: perhaps create an explicit routine `bool useNhwcLayoutForConvOnRocm()` here...see similar comment for batch_norm below

File: tensorflow/core/kernels/conv_grad_input_ops.cc [hunk 4]
```diff
@@ -206,13 +208,26 @@ void LaunchConv2DBackpropInputOp<GPUDevice, T>::operator()(
...
+  TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+      "TF_USE_ROCM_NHWC",
+      /*default_val=*/false, &is_enabled));
+  const bool compute_in_nhwc = DataTypeToEnum<T>::value == DT_HALF &&
+                               (stream->GetGcnArchName().find("gfx908") != std::string::npos ||
+                                stream->GetGcnArchName().find("gfx90a") != std::string::npos)
+                                && is_enabled;
+#else
+  const bool compute_in_nhwc = false;
+#endif
```
Comment (deven-amd) [inline tensorflow/core/kernels/conv_grad_input_ops.cc]: code duplication...can we put this logic in a routine and share it?  see comment below!

File: tensorflow/core/kernels/conv_grad_filter_ops.cc [hunk 5]
```diff
@@ -826,13 +829,26 @@ void LaunchConv2DBackpropFilterOp<Eigen::GpuDevice, T>::operator()(
...
+  TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+      "TF_USE_ROCM_NHWC",
+      /*default_val=*/false, &is_enabled));
+  const bool compute_in_nhwc = DataTypeToEnum<T>::value == DT_HALF &&
+                               (stream->GetGcnArchName().find("gfx908") != std::string::npos ||
+                                stream->GetGcnArchName().find("gfx90a") != std::string::npos) 
+                                && is_enabled;
+#else
+  const bool compute_in_nhwc = false;
+#endif
```
Comment (deven-amd) [inline tensorflow/core/kernels/conv_grad_filter_ops.cc]: code duplication...can we put this logic in a routine and share it?  see comment below!

File: tensorflow/core/kernels/BUILD [hunk 6]
```diff
@@ -4246,7 +4246,7 @@ tf_kernel_library(
         ":redux_functor",
         ":transpose_functor",
         "//tensorflow/core/util:determinism_for_kernels",
-    ] + if_cuda([
+    ] + if_cuda_or_rocm([
```
Comment (deven-amd) [inline tensorflow/core/kernels/BUILD]: is this intentional?

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 7]
```diff
@@ -1331,6 +1333,16 @@ bool BatchnormSpatialPersistentEnabled() {
...
+#elif TF_ROCM_VERSION >= 50000
+//Use the environment variable TF_USE_ROCM_NHWC to force NHWC and NHWC BN
+  static bool is_enabled = [] {
+    bool is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+        "TF_USE_ROCM_NHWC",
+        /*default_val=*/false, &is_enabled));
+    return is_enabled;
+  }();
+  return is_enabled;
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/remapper.cc]: code duplication....see comments below

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc [hunk 8]
```diff
@@ -57,14 +58,23 @@ inline std::pair<int, int> GetNumGPUs(const Cluster& cluster) {
     num_gpus++;
     auto compute_capability_it =
         device.second.environment().find("architecture");
+#if TENSORFLOW_USE_ROCM
+    bool is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+        "TF_USE_ROCM_NHWC",
+        /*default_val=*/false, &is_enabled));
+    if (compute_capability_it->second == "gfx908" || compute_capability_it->second == "gfx90a" && is_enabled) num_volta++;
+#endif
```
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc]: code duplication...see comments below
